discovery_engine/lib/src/domain/search_manager.dart

```diff
@@ -0,0 +1,217 @@
+
+/// Business logic concerning the management of the active search.
+class SearchManager {
+  final Engine _engine;
+  final EventConfig _config;
+  final ActiveSearchRepository _searchRepo;
+  final DocumentRepository _docRepo;
+  final ActiveDocumentDataRepository _activeRepo;
+  final ChangedDocumentRepository _changedRepo;
+  final EngineStateRepository _engineStateRepo;
+
+  SearchManager(
+    this._engine,
+    this._config,
+    this._searchRepo,
+    this._docRepo,
+    this._activeRepo,
+    this._changedRepo,
+    this._engineStateRepo,
+  );
+
+  /// Handle the given search client event.
+  ///
+  /// Fails if [event] does not have a handler implemented.
+  Future<EngineEvent> handleSearchClientEvent(SearchClientEvent event) =>
+      event.maybeWhen(
+        searchRequested: searchRequested,
+        nextSearchBatchRequested: nextSearchBatchRequested,
+        restoreSearchRequested: restoreSearchRequested,
+        searchClosed: searchClosed,
+        orElse: () =>
+            throw UnimplementedError('handler not implemented for $event'),
+      );
+
+  Future<List<api.Document>> _getSearchDocuments(ActiveSearch search) async {
+    final searchDocs =
+        await _engine.getSearchDocuments(search, _config.maxSearchDocs);
+    await _engineStateRepo.save(await _engine.serialize());
+
+    await _docRepo.updateMany(searchDocs.map((e) => e.document));
+    for (final docWithData in searchDocs) {
+      final id = docWithData.document.documentId;
+      await _activeRepo.update(id, docWithData.data);
+    }
+
+    return searchDocs
+        .map((docWithData) => docWithData.document.toApiDocument())
+        .toList();
+  }
+
+  /// Obtain the first batch of search documents and persist to repositories.
+  Future<EngineEvent> searchRequested(
+    String queryTerm,
+    FeedMarket market,
+  ) async {
+    await searchClosed();
+
+    final search = ActiveSearch(
+      queryTerm: queryTerm,
+      requestedPageNb: 1,
+      pageSize: _config.maxSearchDocs,
+      market: market,
+    );
+    final docs = await _getSearchDocuments(search);
+
+    if (docs.isEmpty) {
+      const reason = SearchFailureReason.noResultsAvailable;
+      return const EngineEvent.searchRequestFailed(reason);
+    }
+
+    // TODO: should we persist active search before we get documents?
+    await _searchRepo.save(search);
+
+    return EngineEvent.searchRequestSucceeded(search, docs);
+  }
+
+  /// Obtain the next batch of search documents and persist to repositories.
+  Future<EngineEvent> nextSearchBatchRequested() async {
+    var search = await _searchRepo.getCurrent();
+
+    if (search == null) {
+      const reason = SearchFailureReason.noActiveSearch;
+      return const EngineEvent.nextSearchBatchRequestFailed(reason);
+    }
+
+    // lets update active search params
+    search = search.copyWith(requestedPageNb: search.requestedPageNb + 1);
+    final docs = await _getSearchDocuments(search);
+
+    if (docs.isEmpty) {
```

Comment:
`noResultsAvailable` is not necessary true, we can have a network error or some other kind of error that we cannot necessary observe here. At the moment I would be more to return an empty list.